wandb/sdk/launch/github_reference.py

```diff
@@ -231,3 +238,40 @@ def _update_path(self, dst_dir: str) -> None:
         elif path.is_dir():
             self.directory = self.path
             self.path = None
+
+    def _pull_repo(self, dst_dir: Optional[str] = None) -> None:
+        """Pull the repo to the destination directory, or to a temp directory if not given.
+
+        :returns: the local directory the repo was pulled to
+        """
+        if dst_dir is None and self.local_dir is not None:
+            return
+        import git
+
+        if dst_dir is None:
+            dst_dir = tempfile.TemporaryDirectory().name
+        self.repo_object = git.Repo.clone_from(self.repo_ssh(), dst_dir)
+        self.local_dir = dst_dir
+
+    def commit(self) -> str:
+        """Get git SHA associated with the reference."""
+        self._pull_repo()
+        assert self.repo_object is not None
+        return self.repo_object.head.commit.hexsha  # type: ignore
+
+    def get_requirements(self) -> str:
+        """Pull requirements.txt from the top-level of the repo."""
+        self._pull_repo()
+        with open(f"{self.local_dir}/requirements.txt") as req_file:
+            reqs = req_file.read()
+        return reqs
+
+    def get_python_version(self) -> Optional[str]:
```

Comment:
GitHubReference feels like an odd place for this function to me - it isn't really related to GitHub and I don't see it being used anywhere. Might be better to create a new pyenv.py file?

I also don't think it is correct. The .python-version can contain multiple versions (so that you can use e.g. python 2 and 3 at the same time, or probably more commonly use tox to test a library against a range of python versions)

The pyenv resolution is also more complicated than this with env vars, searching up the directory path, etc.

Reply:
Some context - this PR is pretty much just helper functions to be used for [Griffin's change](https://github.com/wandb/wandb/pull/5661#discussion_r1223301798). So this method (along with `commit()` and `get_requirements()` aren't used yet, but will be when making runless jobs from git repos.

I'm not really sure of a better method to get a python version just from a repo. We can't rely on env vars or anything else related to the executing environment because it's likely that the machine a job will run on is different than the machine the job is being created from. The plan right now is (correct me if I'm wrong @gtarpenning):
1. Check if python version is supplied explicitly as an argument
2. If not, check if `.python-version` is present in the git repo
  2a. Thanks for warning me that `.python-version` can have multiple versions. Probably go with the highest version present in that case?
4. If not, default to the newest version of Python and display a warning

Reply:
I think with pyenv the first version listed will be the default but that would be worth confirming.

Reply:
The [pyenv README](https://github.com/pyenv/pyenv) wasn't very helpful, but after some experimentation it appears the first version listed is what's used. I'll change it to use that, to be consistent with pyenv.

Reply:
I would say for case 3 we should fallback to whatever python version the user is using when creating the job instead of the newest python version

Reply:
Similar comment as above. Return an optional string which is a path to the file, not the content of the file.